src/library/scala/collection/immutable/Map.scala

```diff
@@ -151,10 +151,25 @@ object Map extends MapFactory[Map] {
 
     override def mapFactory: MapFactory[Map] = underlying.mapFactory
 
-    def remove(key: K): WithDefault[K, V] = new WithDefault[K, V](underlying.remove(key), defaultValue)
+    override def + [V1 >: V](kv: (K, V1)): Map.WithDefault[K, V1] = updated(kv._1, kv._2)
 
-    def updated[V1 >: V](key: K, value: V1): WithDefault[K, V1] =
-      new WithDefault[K, V1](underlying.updated(key, value), defaultValue)
+    override def ++ [V2 >: V](xs: collection.Iterable[(K, V2)]): WithDefault[K, V2] = concat(xs)
+
+    override def concat [V2 >: V](xs: collection.Iterable[(K, V2)]): WithDefault[K, V2] = {
+      val concatUnderlying = underlying.concat(xs)
+      if (concatUnderlying.eq(underlying)) this else new WithDefault(concatUnderlying, defaultValue)
+    }
+
+    def remove(key: K): WithDefault[K, V] = {
+      val removedUnderlying = underlying.remove(key)
+      if (removedUnderlying.eq(underlying)) this else new WithDefault[K, V](underlying.remove(key), defaultValue)
```

Comment:
Do you know how much more performant this is compared to the previous version?

Reply:
I am not sure offhand, I can try and benchmark it later.

Reply:
@julienrf Thanks for asking! I confirmed that the new implementation was equal or slower than the old one, so reverted it.